### PR TITLE
Implement value generators

### LINKED ISSUE #37: Generator for test data values of a given type
Should be done after #36.
The task is to implement this function
`def rtypeValueGen[T](t: RType[T]): Gen[T] = ???`

### PR DESCRIPTION
**Pull request by gostkin**
Resolves #37

### CODE REVIEW

File: library/src/test/scala/scalan/RTypeGens.scala [hunk 1]
```diff
@@ -1,15 +1,20 @@
 package scalan
 
+import org.scalacheck.util.Buildable
 import org.scalacheck.{Arbitrary, Gen}
 
+class GenConfiguration() {
+  var maxLength: Int = 100
```
Comment (aslesarenko) [inline library/src/test/scala/scalan/RTypeGens.scala]: Var?
Length of what?

Comment (gostkin) [reply library/src/test/scala/scalan/RTypeGens.scala]: Max length of a generated array

File: library/src/test/scala/scalan/RTypeUtil.scala [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package scalan
+
+import scalan.RType.{ArrayType, PairType, PrimitiveType}
+
+object RTypeUtil {
+  def valueMatchesRType[T](value: T, tA: RType[_]): Boolean = tA match {
+    case prim: PrimitiveType[a] => value match {
+      case b: Byte => true
+      case b: Short => true
```
[the remark this answers is not in the document]
Comment (gostkin) [reply library/src/test/scala/scalan/RTypeUtil.scala]: that was fixed in other pr, now it's moved there

File: library/src/test/scala/scalan/RTypeGens.scala [hunk 3]
```diff
@@ -70,4 +73,32 @@ trait RTypeGens {
...
+    containerOfN[Array, T](count, valGen)
+  }
+
+  def rtypeValueGen[T](t: RType[T], conf: GenConfiguration = new GenConfiguration()): Gen[_] = t match {
+    case prim: PrimitiveType[a] => primitiveValueGen(prim)
+    case arrayType: ArrayType[a] =>
+      getArrayGen(rtypeValueGen(arrayType.tA, conf), conf.maxArrayLength)
+    case pairType: PairType[a, b] =>
+      for { left <- rtypeValueGen(pairType.tFst); right <- rtypeValueGen(pairType.tSnd) } yield (left, right)
+    case (stringType: RType[String]@unchecked) =>
```
Comment (aslesarenko) [inline library/src/test/scala/scalan/RTypeGens.scala]: this matches ANY type, not matched above, not only strings.

File: library/src/test/scala/scalan/RTypeGens.scala [hunk 4]
```diff
@@ -70,4 +73,32 @@ trait RTypeGens {
...
+      Gen.asciiPrintableStr
+    case _ => throw new NotImplementedError("Not supported")
```
Comment (aslesarenko) [inline library/src/test/scala/scalan/RTypeGens.scala]: NotImplementedError is intended for temporary missing part of code. Here you need to raise sys.error, with clear message `s"Cannot create generator for $t"`.
Think about someone getting this message in the log.

File: library/src/test/scala/scalan/RTypeGens.scala [hunk 5]
```diff
@@ -70,4 +73,32 @@ trait RTypeGens {
...
+  }
```
Comment (aslesarenko) [inline library/src/test/scala/scalan/RTypeGens.scala]: @gostkin  Please add CollType support

File: library/src/test/scala/scalan/RTypeTests.scala [hunk 6]
```diff
@@ -23,4 +23,14 @@ class RTypeTests extends PropSpec with PropertyChecks with Matchers with RTypeGe
     }
     typeCoverageChecker.isFullyCovered(typeGenDepth) shouldBe true
   }
+
+  property("RType generate value by type") {
+    import scala.runtime.ScalaRunTime._
+    val minSuccess = MinSuccessful(100)
+    forAll(getBasicTypeGen(typeGenDepth), minSuccess) { t: RType[_] =>
```
Comment (aslesarenko) [inline library/src/test/scala/scalan/RTypeTests.scala]: You should use the same minSuccess value, for which you achieved the full coverage.

File: library/src/test/scala/scalan/RTypeTestUtil.scala [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+      case b: Long => prim == LongType
+      case b: Char => prim == CharType
+      case b: Float => prim == FloatType
+      case b: Double => prim == DoubleType
+      case b: Boolean => prim == BooleanType
+      case b: Unit => prim == UnitType
+      case _ => false
+    }
+    case arrayType: ArrayType[a] => value match {
+      case arr: Array[_] => if (arr.length != 0) valueMatchesRType(arr(0), arrayType.tA) else true
```
Comment (aslesarenko) [inline library/src/test/scala/scalan/RTypeTestUtil.scala]: If you check all the elements then you don't need to deal with the special case.

File: library/src/test/scala/scalan/RTypeGenCoverageChecker.scala [hunk 8]
```diff
@@ -0,0 +1,61 @@
+package scalan
+
+import spire.syntax.all.cfor
+
+trait RTypeGenCoverageChecker {
```
Comment (aslesarenko) [inline library/src/test/scala/scalan/RTypeGenCoverageChecker.scala]: All traits and their methods must be documented.

Please clearly describe: 1) the role of each trait, its responsibility 2) what each abstract method implementation should do in order to comply to the contract represented by the trait.

File: library/src/test/scala/scalan/RTypeGenCoverageChecker.scala [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+  private type TypePosition = (String, Int)
+  private var typePositions: Set[TypePosition] = Set.empty
+
+  override def consider(item: RType[_]): Unit = {
+    decomposeValue(item)
+  }
+
+  override def isFullyCovered(depth: Int): Boolean = {
+    val names = Seq("PrimitiveType", "PairType", "ArrayType", "StringType", "CollType", "ReplCollType", "OptionType")
```
Comment (aslesarenko) [inline library/src/test/scala/scalan/RTypeGenCoverageChecker.scala]: It is better to use Class[_] for classes using `classOf[PairType[_,_]]` for example

File: library/src/test/scala/scalan/RTypeTestUtil.scala [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    case StringType => value match {
+      case str: String => true
+      case _ => false
+    }
+    case collType: CollType[a] => value match {
+      case coll: Coll[_] => coll.tItem == collType.tItem && coll.forall(item => valueMatchesRType(item, collType.tItem))
+      case _ => false
+    }
+    case replCollType: ReplCollType[a] => value match {
+      case coll: Coll[_] => coll.tItem == replCollType.tItem && coll.forall(item => valueMatchesRType(item, replCollType.tItem))
```
Comment (aslesarenko) [inline library/src/test/scala/scalan/RTypeTestUtil.scala]: Shouldn't it be ReplColl here?

File: library/src/test/scala/scalan/RTypeGens.scala (hunk 3 above)
Comment (gostkin) [reply library/src/test/scala/scalan/RTypeGens.scala]: fixed

File: library/src/test/scala/scalan/RTypeGens.scala (hunk 4 above)
Comment (gostkin) [reply library/src/test/scala/scalan/RTypeGens.scala]: fixed

File: library/src/test/scala/scalan/RTypeGens.scala (hunk 5 above)
Comment (gostkin) [reply library/src/test/scala/scalan/RTypeGens.scala]: fixed

File: library/src/test/scala/scalan/RTypeTests.scala (hunk 6 above)
Comment (gostkin) [reply library/src/test/scala/scalan/RTypeTests.scala]: fixed

File: library/src/test/scala/scalan/RTypeTestUtil.scala (hunk 7 above)
Comment (gostkin) [reply library/src/test/scala/scalan/RTypeTestUtil.scala]: fixed

File: library/src/test/scala/scalan/RTypeTestUtil.scala (hunk 10 above)
Comment (gostkin) [reply library/src/test/scala/scalan/RTypeTestUtil.scala]: fixed

File: library/src/test/scala/scalan/RTypeGenCoverageChecker.scala (hunk 9 above)
Comment (gostkin) [reply library/src/test/scala/scalan/RTypeGenCoverageChecker.scala]: fixed

File: library/src/test/scala/scalan/RTypeGenCoverageChecker.scala (hunk 8 above)
Comment (gostkin) [reply library/src/test/scala/scalan/RTypeGenCoverageChecker.scala]: fixed

File: library/src/test/scala/scalan/RTypeGenCoverageChecker.scala [hunk 11]
```diff
@@ -1,28 +1,43 @@
...
 import spire.syntax.all.cfor
 
+/**
+  * In future we could have some other RType generators,
+  * so we will to unify coverage checks through this trait.
+  */
 trait RTypeGenCoverageChecker {
+  /** Take type into consideration.
+    *
+    * @param    item     RType[_] value to be considered
```
Comment (aslesarenko) [inline library/src/test/scala/scalan/RTypeGenCoverageChecker.scala]: @gostkin this comment is useless, it is clear for the signature itself, that the type is RType and the item "will be considered". 
What you need is to describe what does it mean to be "considered". While describing the real meaning you may come to the conclusion that "consider" maybe not the best verb.

File: library/src/test/scala/scalan/RTypeGenCoverageChecker.scala [hunk 12]
```diff
@@ -1,28 +1,43 @@
...
+    */
   def consider(item: RType[_]): Unit
+
+  /** Check if all required types are generated.
+    *
+    * @param    depth     depth of check
```
Comment (aslesarenko) [inline library/src/test/scala/scalan/RTypeGenCoverageChecker.scala]: @gostkin "depth of check" is also useless, please add new information in the comment

File: library/src/test/scala/scalan/RTypeGenCoverageChecker.scala [hunk 13]
```diff
@@ -1,28 +1,43 @@
...
+    */
   def isFullyCovered(depth: Int): Boolean
 }
 
 class FullTypeCoverageChecker extends RTypeGenCoverageChecker {
   import RType._
   import special.collection._
 
-  private type TypePosition = (String, Int)
+  private type TypePosition = (Serializable, Int)
```
Comment (aslesarenko) [inline library/src/test/scala/scalan/RTypeGenCoverageChecker.scala]: Why Serializable, why not Class[_]

File: library/src/test/scala/scalan/RTypeGenCoverageChecker.scala [hunk 14]
```diff
@@ -1,28 +1,43 @@
...
   private var typePositions: Set[TypePosition] = Set.empty
 
   override def consider(item: RType[_]): Unit = {
     decomposeValue(item)
   }
 
   override def isFullyCovered(depth: Int): Boolean = {
-    val names = Seq("PrimitiveType", "PairType", "ArrayType", "StringType", "CollType", "ReplCollType", "OptionType")
+    val names = Seq(classOf[PrimitiveType[_]], classOf[PairType[_, _]], classOf[ArrayType[_]],
```
Comment (aslesarenko) [inline library/src/test/scala/scalan/RTypeGenCoverageChecker.scala]: 'names' doesn't look like adequate name.

File: library/src/test/scala/scalan/RTypeGenCoverageChecker.scala [hunk 15]
```diff
@@ -32,30 +47,30 @@ class FullTypeCoverageChecker extends RTypeGenCoverageChecker {
     true
   }
 
-  private def attachResult(typeName: String, depth: Int) = {
+  private def attachResult(typeName: Serializable, depth: Int) = {
```
Comment (aslesarenko) [inline library/src/test/scala/scalan/RTypeGenCoverageChecker.scala]: use Class[_]

File: library/src/test/scala/scalan/RTypeGenCoverageChecker.scala (hunk 14 above)
Comment (gostkin) [reply library/src/test/scala/scalan/RTypeGenCoverageChecker.scala]: fixed

File: library/src/test/scala/scalan/RTypeGenCoverageChecker.scala (hunk 13 above)
Comment (gostkin) [reply library/src/test/scala/scalan/RTypeGenCoverageChecker.scala]: fixed

File: library/src/test/scala/scalan/RTypeGenCoverageChecker.scala (hunk 12 above)
Comment (gostkin) [reply library/src/test/scala/scalan/RTypeGenCoverageChecker.scala]: fixed

File: library/src/test/scala/scalan/RTypeGenCoverageChecker.scala (hunk 15 above)
Comment (gostkin) [reply library/src/test/scala/scalan/RTypeGenCoverageChecker.scala]: fixed

File: library/src/test/scala/scalan/RTypeGenCoverageChecker.scala (hunk 11 above)
Comment (gostkin) [reply library/src/test/scala/scalan/RTypeGenCoverageChecker.scala]: fixed

**Pull request merged**